docs/readmes/resources/ref_magma_metrics.md

```diff
@@ -0,0 +1,186 @@
+```
+
+Each Python service on the AGW extends the `MagmaService` class which implements
+the Service303 interface. This interface has the
+[`GetMetrics`](https://sourcegraph.com/github.com/magma/magma/-/blob/orc8r/gateway/python/magma/common/service.py?L403:9)
+method which uses the
+[`metrics_export`](https://sourcegraph.com/github.com/magma/magma/-/blob/orc8r/gateway/python/magma/common/metrics_export.py?L22:5)
+module to get metrics from the Python Prometheus client and encode them for
+export over gRPC.
+
+```python
+for metric_family in registry.collect():
+    if metric_family.type in ('counter', 'gauge'):
+        family_proto = encode_counter_gauge(metric_family, timestamp_ms)
+    ...
+```
+
+While each service defines and sets its own metrics, it is up to the `magmad`
+service to collect metrics from all services and export them to the orc8r. When
+the `magmad` service gets started, it reads metrics configuration from
+[magmad.yml](https://sourcegraph.com/github.com/magma/magma/-/blob/feg/gateway/configs/magmad.yml).
+Then, `magmad`schedules its [`MetricsCollector`]() object to collect and upload
+metrics every `metrics_config.sync_interval` seconds. The `MetricsCollector`
+object loops over all Python services and for each, calls `GetMetrics` over gRPC
+to obtain metrics from the Python service. Then, it divides the gRPC structures
+into chunks of 1 MB or less, and uploads these chunks to the `metricsd` orc8r
+service over gRPC through the `Collect` method in `metricsd`.
+
+```python
+chan = ServiceRegistry.get_rpc_channel(
+    'metricsd',
+    ServiceRegistry.CLOUD,
+    grpc_options=self._grpc_options,
+)
+client = MetricsControllerStub(chan)
+...
+sample_chunks = self._chunk_samples(samples)
+for idx, chunk in enumerate(sample_chunks):
+    ...
+    future = client.Collect.future(
+        metrics_container,
+        self.grpc_timeout,
+    )
+```
+
+In **C++** land, each service also exposes the `GetMetrics` method that the
+`magmad` service uses to fetch collected metrics over gRPC. The
+[`MetricsSingleton`](https://sourcegraph.com/github.com/magma/magma/-/blob/orc8r/gateway/c/common/service303/MetricsSingleton.cpp)
+class provides helpers and wrappers around the Prometheus C++ client.
+[`service303.cpp`](https://sourcegraph.com/github.com/magma/magma/-/blob/lte/gateway/c/core/oai/tasks/service303/service303.cpp)
+further wraps `MetricsSingleton` methods to make it even more convenient for
+Magma services to set and upload Prometheus metrics.
+
+### Phase 2: Orchestrator and Prometheus Edge Hub
+
+The `metricsd` cloud service reads metric names and labels from enums defined in
+[`metricsd.proto`](https://sourcegraph.com/github.com/magma/magma/-/blob/orc8r/protos/metricsd.proto)
+and accepts gRPC messages accordingly. Once it has received metrics from the
+AGW, it pushes them again over gRPC to the
+[`MetricsExporter`](https://sourcegraph.com/github.com/magma/magma/-/blob/orc8r/cloud/go/services/metricsd/protos/exporter.proto)
```

Comment:
Consider noting that this is an extension-based architecture, such that anyone can extend Orc8r with their own metrics sink

https://magma.github.io/magma/docs/next/orc8r/architecture_modularity